lib/concurrent/synchronization/java_pure_object.rb

```diff
@@ -32,6 +33,10 @@ def ns_signal
           JRuby.reference0(self).notify
           self
         end
+
+        def ensure_ivar_visibility!
+          # relying on undocumented behavior of JRuby, ivar access is volatile
```

Comment:
The reified variable logic is experimental and not enabled by default, but if we went forward with it we'd need to address the volatility difference.

Normally ivars are stored in an array, and opaque array accesses (same code, many different arrays) don't fold away on JVM because of the boundschecking and indirection involved. Updates of that array are done in a volatile way, as @chrisseaton mentioned. But writes *into* the array also go through the fence/volatile logic @pitr-ch mentioned, so all writes should be volatile. And I *believe* that traversing the volatile ivar array reference for reads forces happens-before visibility on the values (since the JVM can't optimistically optimize the access into a non-volatile cache hit).

We have not made this official for exactly the reason @pitr-ch mentions: no other implementation makes a hard volatility guarantee and we want to be able to pull back a bit on volatility later.

As for performance concerns...JRuby's current ivar strategy already injects so much overhead, making them volatile-by-default may not even be noticeable (we could test). But saying we're always volatile now means we can't improve ivar performance by removing volatility later.

It's a conundrum, to be sure.